### ISSUE TITLE
Request authority Wikidata be supported

### ISSUE THREAD

**Issue by elrayle** [LD4P/qa_server#15]
**Authority:** [Wikidata](https://www.wikidata.org/wiki/Wikidata:Data_access)
**Download:** https://www.wikidata.org/wiki/Wikidata:Database_download

**Comment by elrayle** [comment]
This request will not be immediately acted upon.  Research is required to determine if we can cache data, whether we want to, is direct access a possibility?  There are also questions about what data from wikidata we want to query and fetch.

**Comment by eichmann** [comment]
I've refreshed the existing triplestore with current data as of 11/13, so yes, caching is possible.

**Comment by sfolsom** [comment]
The conclusion if a conversation with @elrayle and others during the Cornell meeting today is that we need subauthorities based on the spreadsheet for Wikidata column A, https://docs.google.com/spreadsheets/d/1rPvEoP9iYNkxJ0eWC8gXe3ci7e6mhW0da59xkGhadi0/edit#gid=965082913 (The spreadsheet isn't ready for action yet)

**Comment by eslao** [comment]
@elrayle @eichmann I've put together some notes in the hopes that we can revisit this request, with an eye toward first implementing a "generic" lookup for any type of Wikidata entity: https://docs.google.com/document/d/1zHfziWP2I9lfNnrrVVX2Xyvim5cXwtTRU8GddKCbJy4/edit#

We've split the Wikidata tab in the spreadsheet into two tabs -- one for the "generic" lookup, and another for potential subauthority-esque lookups: https://docs.google.com/spreadsheets/d/1rPvEoP9iYNkxJ0eWC8gXe3ci7e6mhW0da59xkGhadi0/edit#gid=965082913

Please let me know if this is enough to move forward, and/or if a call (with me and/or @mmcgee and/or @sfolsom) would be helpful. Thanks!

**Comment by zimeon** [comment]
Suggest trying to implement the generic wikidata query outlined in the spreadsheet as a direct SPARQL query against wikidata -- is that realistic? is that fast enough?

**Comment by sfolsom** [comment]
If there is concern about indexing all of Wikidata for reasons of scale and keeping the data current, we may want to look into the API @eslao documented in the Google doc above. Wbsearchentities module (gets item numbers, labels, and description string by label string): https://www.wikidata.org/w/api.php?action=help&modules=wbsearchentities

**Comment by zimeon** [comment]
Some possibly useful starting info on search by label in https://stackoverflow.com/questions/38527828/how-to-query-wikidata-items-using-its-labels

**Comment by eslao** [comment]
@zimeon Thanks for finding that. Would a direct SPARQL query on Wikidata, or the API that @sfolsom mentions, enable us to create a lookup that meets the specs that @mmcgee and I have laid out? As I said before, I'd be happy to set up a call to weigh our options, if there are trade-offs between the possible approaches.

**Comment by elrayle** [comment]
@eslao QA has not yet worked with SPARQL queries.  I plan to use the wikidata as a first step of getting QA to play nicely with QA.

QA is approaching a major release to add context.  I plan to start this work once that is complete.

One thing that can speed up this work is an example SPARQL query in the structure desired to get the data you want.  If someone has time to do that, it will make the QA integration easier.  The query can be tested at... https://query.wikidata.org/

**Comment by eslao** [comment]
@elrayle I've started by putting the beginnings of a query here: https://github.com/LD4P/Wikidata/blob/master/generic%20entity/wikidata_generic_entity.rq (@mmcgee and I will be adding more of these in the same folder)

The query I started with is looking for an exact string match; I started to experiment with a (contains()) filter (as many labels and aliases in Wikidata include definite articles and whatnot) --  this seems to time out, but I'll keep playing with it. We may need to compare performance between a few different queries. Should I be exploring regex options for non-exact string matches, or does QA do anything to support fuzzy matching that might apply here?

**Comment by eichmann** [comment]
Here's a list of class instance counts from a just-completed triplestore load:

6130 owl:Class ;
19579 owl:DatatypeProperty .
46941 owl:ObjectProperty .
6130 owl:Restriction ;
68362885 schema:Article ;
58935966 schema:Dataset ;
 312 schema:Dataset,
7423600 wikibase:GlobecoordinateValue ;
26226739 wikibase:Item .
30495072 wikibase:Item ;
6130 wikibase:Property,
3996256 wikibase:QuantityValue ;
51598093 wikibase:Reference ;
  89 wikibase:Reference,
1209842 wikibase:Statement ;
716414823 wikibase:Statement,
5805164 wikibase:TimeValue ;

**Comment by elrayle** [comment]
@eslao I wanted to touch base with you on your needs for wikidata.  

### Current Status in QA

At this point, I have added the ability to search wikidata API.  It does not have a lot of extended context because they do not return much information.

Example search query:  https://lookup.ld4l.org/authorities/search/wikidata/item?q=milk

This is fast, but doesn't have much information.  And there is no way to limit results to entities or other additional filters on searching.

Example fetch single term:  https://lookup.ld4l.org/authorities/fetch/linked_data/wikidata_direct?format=n3&uri=http%3A%2F%2Fwww%2Ewikidata%2Eorg%2Fentity%2FQ8495

This is very slow because they return Gbytes of data when dereferencing a term.

### Current Status in Sinopia

This is not deployed to Sinopia, but I can add this if it is sufficient, at least for now.

### Requested Feedback

How are you planning to use wikidata in Sinopia?  Options...
* Search for a result to include a label and URI in a Sinopia form field
* Fetch a wikidata entity to use multiple triples for filling in multiple fields
* Combination of both

Is the current information adequate for your needs?

**Comment by elrayle** [comment]
Based on discussions in the QA Study Hall focused on wikidata, there was general agreement that the search is adequate.  It would be nice to have extended context at some point. There is a PR to add wikidata search to Sinopio.  (PR https://github.com/LD4P/sinopia_editor/pull/1662)

Term fetch, which dereferences the URI, is too slow to be useful.  Possible solutions...

OPTION 1: Explore the many wikidata APIs to see if one will return less data for a single term and be more performant.
OPTION 2: Explore the ability to use SPARQL query to fetch a limited set of data for a single term. 
 Requires update to QA to support SPARQL.

**Comment by sfolsom** [comment]
A first pass has been made on this issue using the https://www.wikidata.org/w/api.php?action=help&modules=wbsearchentities API. After some experience with this approach we can decide if it's meeting our needs, or if we need to come up with an approach that provides more context.
